autograph integration tests broken

As of this morning. Previously green tests are busted.

refs: https://travis-ci.org/mozilla-releng/signingscript/builds/438798351

```
signingscript/test/integration/test_autograph.py::test_integration_autograph_apk FAILED
=================================== FAILURES ===================================
________________________ test_integration_autograph_apk ________________________
context = <scriptworker.context.Context object at 0x7f23af00f940>
tmpdir = local('/tmp/pytest-of-travis/pytest-0/test_integration_autograph_apk0')
    @pytest.mark.asyncio
    @skip_when_no_autograph_server
    async def test_integration_autograph_apk(context, tmpdir):
        file_name = 'app.apk'
        original_file_path = os.path.join(TEST_DATA_DIR, file_name)
        copied_file_folder = os.path.join(context.config['work_dir'], 'cot', 'upstream-task-id1')
        makedirs(copied_file_folder)
        shutil.copy(original_file_path, copied_file_folder)
    
        context.config['signing_server_config'] = _write_server_config(tmpdir)
        context.task = _craft_task([file_name], signing_format='autograph_apk')
    
        keystore_path = os.path.join(tmpdir, 'keystore')
        certificate_path = os.path.join(TEST_DATA_DIR, 'autograph_apk.pub')
        certificate_alias = 'autograph_apk'
        _instanciate_keystore(keystore_path, certificate_path, certificate_alias)
    
        await async_main(context)
    
        signed_path = os.path.join(tmpdir, 'artifact', file_name)
>       assert _verify_apk_signature(keystore_path, signed_path, certificate_alias)
E       AssertionError: assert False
E        +  where False = _verify_apk_signature('/tmp/pytest-of-travis/pytest-0/test_integration_autograph_apk0/keystore', '/tmp/pytest-of-travis/pytest-0/test_integration_autograph_apk0/artifact/app.apk', 'autograph_apk')
signingscript/test/integration/test_autograph.py:207: AssertionError
```

stdout and stderr from `_verify_apk_signature` shelling out:

```
- Signed by "CN=Signingscript Test Key, OU=Unknown, O=Unknown, L=Unknown, ST=Unknown, C=Unknown"                                                        
    Digest algorithm: SHA-256                                                                                                                                                                 
    Signature algorithm: SHA256withSHA256withRSA, 2048-bit key                                                           
                                                                                                                                                                                              jar verified, with signer errors.                                                                                                                               
                                                                                                         
Error:                                                                                                    
This jar contains entries whose certificate chain is invalid. Reason: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certificatio
n path to requested target                                                                                   
This jar contains signed entries which are not signed by the specified alias(es).                         
This jar contains signed entries that are not signed by alias in this keystore.
This jar contains entries whose signer certificate is self-signed.

Warning:
This jar contains signatures that do not include a timestamp. Without a timestamp, users may not be able to validate this jar after any of the signer certificates expire (as early as 2046-01
-05).
```

Since this only breaks in 3.7 it's probably due to a Python change. There are a few subprocess changes in 3.7, but nothing obvious.

> bpo-32844: Fix wrong redirection of a low descriptor (0 or 1) to stderr in subprocess if another low descriptor is closed.
> bpo-31178: Fix string concatenation bug in rare error path in the subprocess module

https://docs.python.org/3/whatsnew/changelog.html

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,16 +4,29 @@ matrix:
   fast_finish: true
   include:
     - python: 3.6
+      dist: xenial
       env:
         - TOXENV=py36
         - AUTOGRAPH_INTEGRATION=1
+        - JAVA_HOME=/usr/lib/jvm/java-8-oracle
+      sudo: true
     - python: 3.7
       dist: xenial
       env:
         - TOXENV=py37
         - AUTOGRAPH_INTEGRATION=1
+        - JAVA_HOME=/usr/lib/jvm/java-8-oracle
       sudo: true
 
+before_install:
+    # Install oracle java 8, because jarsigner in openjdk 11.0.1+13 breaks
+    # integration tests.
+    # https://github.com/mozilla-releng/signingscript/issues/80
+    - sudo add-apt-repository -y ppa:webupd8team/java
+    - sudo apt-get update
+    - echo debconf shared/accepted-oracle-license-v1-1 select true | sudo debconf-set-selections
+    - echo debconf shared/accepted-oracle-license-v1-1 seen true | sudo debconf-set-selections
+    - sudo apt-get install oracle-java8-installer
 install:
     - travis_retry pip install tox
 script:

--- a/signingscript/sign.py
+++ b/signingscript/sign.py
@@ -541,7 +541,7 @@ async def _convert_dmg_to_tar_gz(context, from_):
     work_dir = context.config['work_dir']
     abs_from = os.path.join(work_dir, from_)
     # replace .dmg suffix with .tar.gz (case insensitive)
-    to = re.sub('\.dmg$', '.tar.gz', from_, flags=re.I)
+    to = re.sub(r'\.dmg$', '.tar.gz', from_, flags=re.I)
     abs_to = os.path.join(work_dir, to)
     dmg_executable_location = context.config['dmg']
     hfsplus_executable_location = context.config['hfsplus']

--- a/signingscript/test/integration/test_autograph.py
+++ b/signingscript/test/integration/test_autograph.py
@@ -2,6 +2,7 @@ import aiohttp
 import asyncio
 import copy
 import json
+import logging
 import os
 import pytest
 import subprocess
@@ -20,6 +21,7 @@ from signingscript.test.integration import skip_when_no_autograph_server
 assert context  # silence flake8
 
 
+log = logging.getLogger(__name__)
 DATA_DIR = os.path.join(os.path.dirname(__file__), '..', '..', 'data')
 TEST_DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')
 
@@ -165,22 +167,36 @@ async def test_integration_autograph_mar_sign_hash(context, tmpdir):
         assert do_verify(signed_path, keyfiles=[mar_pub_key_path]), "Mar signature doesn't match expected key"
 
 
+def _get_java_path(tool_name):
+    if os.environ.get('JAVA_HOME'):
+        return os.path.join(os.environ['JAVA_HOME'], 'bin', tool_name)
+    return tool_name
+
+
 def _instanciate_keystore(keystore_path, certificate_path, certificate_alias):
     keystore_password = '12345678'
-    subprocess.run([
-        'keytool', '-import', '-noprompt',
+    cmd = [
+        _get_java_path('keytool'), '-import', '-noprompt',
         '-keystore', keystore_path, '-storepass', keystore_password,
         '-file', certificate_path, '-alias', certificate_alias
-    ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+    ]
+    log.info("running {}".format(cmd))
+    subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True
+    )
 
 
 def _verify_apk_signature(keystore_path, apk_path, certificate_alias):
-    command = subprocess.run([
-        'jarsigner', '-verify', '-strict', '-verbose',
+    cmd = [
+        _get_java_path('jarsigner'), '-verify', '-strict', '-verbose',
         '-keystore', keystore_path,
         apk_path,
         certificate_alias
-    ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+    ]
+    log.info("running {}".format(cmd))
+    command = subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True
+    )
     return command.returncode == 0
 
 

--- a/tox.ini
+++ b/tox.ini
@@ -11,6 +11,7 @@ passenv =
     TRAVIS_JOB_ID
     TRAVIS_BRANCH
     AUTOGRAPH_INTEGRATION
+    JAVA_HOME
 
 deps =
     coverage>=4.2b1

